### ISSUE TITLE
Craft CMS update from 3.6.17 to 3.7 causes db migration issue

### ISSUE THREAD

**Issue by ryguy2407** [putyourlightson/craft-blitz#332]
I'm trying to update my Craft CMS site from 3.6.17 to 3.7 but I get a database migration error.

I run the updates from the control panel but get the following stack trace when Craft tries to run its migrations:

```
Database Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'paginate' in 'field list'
The SQL being executed was: SELECT uri, paginate
FROM blitz_caches
WHERE (siteId=1) AND (NOT (paginate IS NULL))

Migration: craft\migrations\m210613_145522_sortable_global_sets

Output:

> add column sortOrder smallint UNSIGNED AFTER fieldLayoutId to table {{%globalsets}} ... done (time: 0.140s)
> create index idx_nepomyyaszikrdvtwlctwsbpdkeulapfocpv on {{%globalsets}} (sortOrder) ... done (time: 0.057s)
Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'paginate' in 'field list'
The SQL being executed was: SELECT uri, paginate
FROM blitz_caches
WHERE (siteId=1) AND (NOT (paginate IS NULL)) (/home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/Schema.php:678)
#0 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/Command.php(1307): yii\db\Schema->convertException()
#1 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/Command.php(1168): yii\db\Command->internalExecute()
#2 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/Command.php(410): yii\db\Command->queryInternal()
#3 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/Query.php(249): yii\db\Command->queryAll()
#4 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/db/ActiveQuery.php(133): yii\db\Query->all()
#5 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/helpers/SiteUriHelper.php(152): yii\db\ActiveQuery->all()
#6 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/helpers/SiteUriHelper.php(93): putyourlightson\blitz\helpers\SiteUriHelper::getPaginatedUrisForSite()
#7 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/helpers/SiteUriHelper.php(59): putyourlightson\blitz\helpers\SiteUriHelper::getSiteUrisForSite()
#8 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/services/RefreshCacheService.php(465): putyourlightson\blitz\helpers\SiteUriHelper::getAllSiteUris()
#9 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/services/RefreshCacheService.php(198): putyourlightson\blitz\services\RefreshCacheService->refreshAll()
#10 /home/vagrant/craft/bnphn/vendor/putyourlightson/craft-blitz/src/Blitz.php(327): putyourlightson\blitz\services\RefreshCacheService->addElement()
#11 [internal function]: putyourlightson\blitz\Blitz->putyourlightson\blitz\{closure}()
#12 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Event.php(312): call_user_func()
#13 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Component.php(637): yii\base\Event::trigger()
#14 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/Elements.php(2773): yii\base\Component->trigger()
#15 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/Elements.php(773): craft\services\Elements->_saveElementInternal()
#16 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/Globals.php(383): craft\services\Elements->saveElement()
#17 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/base/ApplicationTrait.php(1668): craft\services\Globals->handleChangedGlobalSet()
#18 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/ProjectConfig.php(1189): craft\web\Application->craft\base\{closure}()
#19 [internal function]: craft\services\ProjectConfig->handleChangeEvent()
#20 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Component.php(628): call_user_func()
#21 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/ProjectConfig.php(752): yii\base\Component->trigger()
#22 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/ProjectConfig.php(1179): craft\services\ProjectConfig->_processConfigChangesInternal()
#23 [internal function]: craft\services\ProjectConfig->handleChangeEvent()
#24 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Component.php(628): call_user_func()
#25 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/ProjectConfig.php(749): yii\base\Component->trigger()
#26 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/ProjectConfig.php(521): craft\services\ProjectConfig->_processConfigChangesInternal()
#27 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/migrations/m210613_145522_sortable_global_sets.php(38): craft\services\ProjectConfig->set()
#28 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/db/Migration.php(52): craft\migrations\m210613_145522_sortable_global_sets->safeUp()
#29 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/db/MigrationManager.php(232): craft\db\Migration->up()
#30 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/db/MigrationManager.php(148): craft\db\MigrationManager->migrateUp()
#31 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/services/Updates.php(250): craft\db\MigrationManager->up()
#32 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/controllers/BaseUpdaterController.php(509): craft\services\Updates->runMigrations()
#33 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/controllers/UpdaterController.php(203): craft\controllers\BaseUpdaterController->runMigrations()
#34 [internal function]: craft\controllers\UpdaterController->actionMigrate()
#35 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#36 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams()
#37 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/web/Controller.php(190): yii\base\Controller->runAction()
#38 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Module.php(534): craft\web\Controller->runAction()
#39 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/web/Application.php(271): yii\base\Module->runAction()
#40 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/web/Application.php(672): craft\web\Application->runAction()
#41 /home/vagrant/craft/bnphn/vendor/craftcms/cms/src/web/Application.php(208): craft\web\Application->_processUpdateLogic()
#42 /home/vagrant/craft/bnphn/vendor/yiisoft/yii2/base/Application.php(392): craft\web\Application->handleRequest()
#43 /home/vagrant/craft/bnphn/web/index.php(21): yii\base\Application->run()
#44 {main}
```
I'm working with P&T support as well and they released a patch to address migration issues, but I'm still getting this error.

Running Blitz 3.9.0

**Comment by bencroker** [comment]
Hmm, the `paginate` column was added in Blitz 3.10.0. Can you try updating Blitz to the latest version before you try updating Craft again?

**Comment by jsmrtn** [comment]
Can confirm the migration completes successfully by updating Blitz to 3.10 before pushing to Craft 3.7

**Comment by bencroker** [comment]
Great!

**Comment by brandonkelly** [comment]
Hey @bencroker, this crossed our radar. My understanding is that the issues is that a global set is saved during the 3.7 migrations, which triggers Blitz caching _before_ Blitz’ own migrations have run.

This is actually a fairly common situation which Blitz should be guarding against, by checking its own installed schema version and compensating when the migration hasn’t been run yet. Something like:

```php
$schemaVersion = Craft::$app->projectConfig->get('plugins.blitz.schemaVersion');
if (version_compare($schemaVersion, '3.10.0', '>')) {
    $query
        ->addSelect('paginate');
        ->andWhere(['not', ['paginate' => null]]);
}
```

Craft does this in quite a few places as well, such as:

https://github.com/craftcms/cms/blob/0eff1aadb1110f0f35c92b6d8c86a0740c34a5fc/src/elements/db/GlobalSetQuery.php#L131-L134

**Comment by khalwat** [comment]
@bencroker I do something similar to ensure that all migrations have run: 

https://github.com/nystudio107/craft-seomatic/blob/v3/src/Seomatic.php#L461

**Comment by bencroker** [comment]
Thanks @brandonkelly (and @khalwat), I wasn't aware of such cases already existing in Craft.

Have fixed in https://github.com/putyourlightson/craft-blitz/commit/a4f82f48ea7ea888af5d1665de33a56e5530d550 and released in version 3.10.1.
